src/overlays/actors/ovl_En_Slime/z_en_slime.c

```diff
@@ -88,103 +128,1169 @@ static DamageTable D_80A31B1C = {
+    this->iceBlockSnapPos.z = EnSlime_SnapIceBlockPosition(this->actor.world.pos.z, this->actor.home.pos.z);
+    Actor_SetColorFilter(&this->actor, 0x8000, 0x80FF, 0x2000, 10);
+    this->eyeTexIndex = EN_SLIME_EYETEX_OPEN;
+    this->iceBlockTimer = 254;
+    this->actionFunc = EnSlime_SpawnIceBlock;
+}
+
+/**
+ * Attempt to move the chuchu to snapped position and
+ * spawn an ice block actor as a child of the chuchu actor.
+ * If ice block spawn fails, chuchu returns to idle.
+ */
+void EnSlime_SpawnIceBlock(EnSlime* this, PlayState* play) {
+    s32 stepCheck; // bool
+
+    stepCheck = Math_StepToF(&this->actor.world.pos.x, this->iceBlockSnapPos.x, 10.0f);
+    stepCheck &= Math_StepToF(&this->actor.world.pos.z, this->iceBlockSnapPos.z, 10.0f);
+
+    this->actor.colorFilterTimer = 10;
+    if (stepCheck) {
+        this->actor.child =
+            Actor_SpawnAsChild(&play->actorCtx, &this->actor, play, ACTOR_OBJ_ICEBLOCK, this->actor.world.pos.x,
+                               this->actor.world.pos.y + 30.0f, this->actor.world.pos.z, 0, 0, 0, 0);
+        if (this->actor.child != NULL) {
+            this->actor.child->cutscene = this->actor.cutscene;
+            EnSlime_SetIceBlockedState(this);
+        } else {
+            this->actor.colorFilterTimer = 0;
+            this->collider.base.acFlags |= AC_ON;
+            this->iceBlockTimer = 255;
+            this->actor.flags |= ACTOR_FLAG_1;
+            this->actor.gravity = -2.0f;
+            EnSlime_ReturnToIdle(this);
+        }
+    }
+}
+
+/**
+ * Enter state as a chuchu frozen in an ice block.
+ * This function is called after ice block spawn is successful.
+ * Action function set to UpdateIceBlockedState
+ */
+void EnSlime_SetIceBlockedState(EnSlime* this) {
+    this->actor.flags |= ACTOR_FLAG_10;
+    this->actionFunc = EnSlime_UpdateIceBlockedState;
+}
+
+/**
+ * Decrement the ice block timer, or if it has hit zero, check to
+ * see if the ice block child actor link has been broken or if it has
+ * shrunk sufficiently.
+ * If timer is up and ice block is sufficiently "gone", init unfreeze.
+ */
+void EnSlime_UpdateIceBlockedState(EnSlime* this, PlayState* play) {
+    if (this->iceBlockTimer == 0) {
+        // Proceed to next action if ice block link or its update link are broken or the ice block has shrunk
+        // sufficiently.
+        if ((this->actor.child == NULL) || (this->actor.child->update == NULL) ||
+            !(this->actor.child->scale.y >= 0.1f)) {
+            EnSlime_InitIceBlockUnfreeze(this);
+        }
+    } else {
+        this->actor.colorFilterTimer = 10;
+        if ((this->iceBlockTimer - 5) < 0) {
+            this->iceBlockTimer = 0;
+        } else {
+            this->iceBlockTimer -= 5;
+        }
+    }
+}
+
+/**
+ * Prepare the actor to behave as if stunned. Set speed to zero,
+ * and set action function to UpdateStunnedState.
+ */
+void EnSlime_EnterStunnedState(EnSlime* this) {
```

Comment:
Another setup function; we typically call this one something like `EnSlime_SetupStunned` and the one below `EnSlime_Stunned`. Though we are not consistent; sometimes we use "Stun" instead of "Stunned".